mm/kfence/static_branch.c

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/atomic.h>
+#include <linux/mm.h>
+#include <linux/static_key.h>
+#include <linux/wait.h>
+#include <linux/workqueue.h>
+
+#include "kfence.h"
+#include "../slab.h"
+
+/* The static key to set up a KFENCE allocation. */
+DEFINE_STATIC_KEY_FALSE(kfence_allocation_key);
+
+/* Gates the allocation, ensuring only one succeeds in a given period. */
+static atomic_t allocation_gate = ATOMIC_INIT(1);
+/* Wait queue to wake up heartbeat timer task. */
+static DECLARE_WAIT_QUEUE_HEAD(allocation_wait);
+
+void *kfence_alloc_with_size(struct kmem_cache *s, size_t size, gfp_t flags)
+{
+	void *ret;
+
+	if (atomic_inc_return(&allocation_gate) > 1)
+		return NULL;
+	wake_up(&allocation_wait);
+
+	if (!kfence_is_enabled())
+		return NULL;
+
+	if ((size > PAGE_SIZE) || (s->size > PAGE_SIZE))
+		return NULL;
+	if (s->ctor || (s->flags & SLAB_TYPESAFE_BY_RCU))
+		return NULL;
+
+	ret = kfence_guarded_alloc(s, size, flags);
+
+	return ret;
+}
+EXPORT_SYMBOL(kfence_alloc_with_size);
```

Comment:
Does this need to be exported? I think it's only called from slub/slab?

Reply:
No doesn't. Inherited from naive.c, which currently has same.

My guess is we should only keep one implementation for the final version, and this file will merge back with core.c. When that happens, we'll scrutinize every line of code. Until then, things look a little "rough".

Reply:
`kfence_alloc_with_size()` used to be called from `kmalloc()`, which is an always_inline function.
Marco has removed the call in e66bf3b4d652d7e84a009bbe20ccfdbfaf0e33b1, so we don't need to export it anymore.